src/memory_manager.rs

```diff
@@ -103,12 +106,14 @@ pub fn bind_mutator<VM: VMBinding>(
     mutator
 }
 
-/// Reclaim a mutator that is no longer needed.
+/// Report to MMTk that a mutator that is no longer needed. A binding should not attempt
+/// to use the mutator after this call. MMTk will not attempt to reclaim the memory for the
+/// mutator, so a binding should properly reclaim the memory for the mutator after this call.
 ///
 /// Arguments:
 /// * `mutator`: A reference to the mutator to be destroyed.
-pub fn destroy_mutator<VM: VMBinding>(mutator: Box<Mutator<VM>>) {
-    drop(mutator);
+pub fn destroy_mutator<VM: VMBinding>(mutator: &mut Mutator<VM>) {
+    mutator.on_destroy();
```

Comment:
Does this imply that bindings should be responsible for releasing `Box<Mutator<VM>>`?

Reply:
Yes, the binding should be responsible for releasing. That is made clear in the documentation in this PR.

This change is necessary. Our current API (which requires `Box<Mutator>`) does not really work at all in some cases. For example, a binding copies the mutator content to their thread local storage. When they call `destroy_mutator()`, they pass their thread local storage pointer as `Box<Mutator>`, but Rust is not able to drop the `Box` as it is not the original memory Rust allocates.

Reply:
> I have a different concern. If the API is symmetric, their behaviours should be symmetric, too. If `create_mutator` allocates a box and returns a `Box<Mutator<VM>>`, then `destroy_mutator` should take a `Box<Mutator<VM>>` and destroy the box. That's intuitive. Given that we want the mutator to be created and destroyed differently, I suggest we change the API so the creation and the destruction look different.

If you look at `bind_mutator` and `destroy_mutator`, yes, I think you are right. It is intuitive that 'destroy' would undo whatever you have done in 'bind'. However, we currently have `bind_mutator() -> Box<Mutator>`, `destroy_mutator(&mut Mutator)` and `flush_mutator(&mut Mutator)`. I feel it is more or less okay -- you create the mutator first, then do different things with it. The documentation makes it clear what each method expects, and how a binding can manage the `Box`'s lifetime. I feel it is okay. Let me know what you think.

> Creation (`bind_mutator`) can remain like this, or change it to a method of `MMTK`: `mmtk.create_mutator(tls)`.
> 
> I suggest expose `Mutator::on_destroy(&mut self)` (or rename it to simply `destroy()`) as a public method. When the VM calls this method on the `Mutator` in its own data structure, it is clear that it is not the same as the creation API.
> 
> p.s. This is a chance to gradually move away from the `memory_manager` module. See #658

`Mutator::on_destroy()` is public.

Reply:
Yes.  It is OK to "create it first and then do different things with it".  I think eventually VMs will call `mutator.destroy()` and `mutator.flush()` directly and not call top-level functions in `memory_manager`, but we can leave that refactoring to another day.